### ISSUE TITLE
XLA:CPU segfault in YNN/slinky fusion for rank-7 f32 reduce with a degenerate dim (jaxlib 0.10.2 regression from 0.10.1)

### ISSUE THREAD

**Issue by rogeriojorge** [jax-ml/jax#40412]
# XLA:CPU segfault in YNN/slinky fusion for rank-7 f32 reduce with a degenerate dim (jaxlib 0.10.2, regression from 0.10.1)

## Description

On the CPU backend, jaxlib 0.10.2 segfaults while **executing** an f32
elementwise-multiply + reduce fusion whose operands have **rank 7** and contain a
**size-1 dimension that is not the reduced axis**. jaxlib 0.10.1 runs the same
program correctly, as does float64 on 0.10.2.

The fault is asynchronous (it happens on an Eigen worker thread inside the thunk
executor), so the Python-level traceback lands wherever the main thread happened
to be — compiling a later module, hashing a sharding, materialising an array.
That makes it look like several unrelated bugs until you get a native backtrace.

## Reproducer

No FFT, no autodiff, no third-party code:

```python
import jax, jax.numpy as jnp

S = (3, 1, 2, 2, 8, 8, 8)
u = jax.random.normal(jax.random.PRNGKey(0), S, dtype=jnp.float32)
v = jax.random.normal(jax.random.PRNGKey(1), S, dtype=jnp.float32)
jax.jit(lambda a, b: jnp.sum(a * b, axis=3))(u, v)   # SIGSEGV
```

```
JAX_PLATFORMS=cpu python repro.py
Segmentation fault (core dumped)
```

## What varies

| case | result |
|---|---|
| `(3, 1, 2, 2, 8, 8, 8)`, reduce axis 3, f32 | **SIGSEGV** |
| `(1, 3, 2, 2, 8, 8, 8)`, reduce axis 3, f32 | **SIGSEGV** |
| `(3, 2, 2, 2, 8, 8, 8)`, reduce axis 3, f32 (no size-1 dim) | ok |
| `(3, 1, 2, 2, 8, 8, 8)`, reduce axis 1 (reduce *is* the size-1 dim), f32 | ok |
| `(3, 2, 2, 8, 8, 8)`, reduce axis 2, f32 (rank 6) | ok |
| `(2, 1, 2, 2, 2, 2, 2)`, reduce axis 3, f32 (rank 7, small) | ok — fusion not routed to YNN |
| `(3, 1, 2, 2, 8, 8, 8)`, reduce axis 3, **f64** | ok — YNN path is f32-only |
| same program, jaxlib **0.10.1** | ok |
| same program, GPU backend | ok |

So the trigger is: f32, rank 7, a degenerate dimension other than the reduced
one, and enough work that XLA routes the fusion to YNN.

## Native backtrace

```
slinky::internal::can_fuse<slinky::buffer<void const, 8ul>>(...)      <-- SIGSEGV
slinky::fuse_contiguous_dims<slinky::buffer<void const, 8ul>>(...)      static_transpose.cc
ynn::(anonymous)::make_transpose_impl(...)::$_0                         static_transpose.cc
slinky::(anonymous)::evaluator::eval_non_inlined(slinky::stmt const&)   evaluate.cc
slinky::(anonymous)::evaluator::eval(slinky::allocate const*)           evaluate.cc
slinky::evaluate(slinky::stmt const&, slinky::eval_context&)            evaluate.cc
ynn_invoke_runtime                                                      runtime.cc
xla::cpu::YnnFusionThunk::YnnExecutable::Invoke(...)                    ynn_fusion_thunk.cc
xla::cpu::YnnFusionThunk::Execute(...)                                  ynn_fusion_thunk.cc
xla::cpu::ThunkExecutor::TracedExecute(...)                             thunk_executor.cc
xla::cpu::ThunkExecutor::Execute<FifoReadyQueue>(...)                   thunk_executor.cc
Eigen::ThreadPoolTempl<tsl::thread::EigenEnvironment>::WorkerLoop(int)  threadpool.cc
```

Fault address is `0x3ffc7` — a small non-null offset, i.e. a bad pointer rather
than a stack overflow. The faulting thread's stack region was
`[0x776ba5ffd000, 0x776ba67fd000)` (8 MB), nowhere near the fault address, and
raising `ulimit -s` to 512 MB or running the computation from a 512 MB-stack
thread does not help.

Note `slinky::buffer<void const, 8ul>` — an 8-slot static dim buffer — being used
for a rank-7 fusion whose output XLA bitcasts to rank 8. `fuse_contiguous_dims`
collapsing the degenerate dimension looks like the relevant ingredient.

## Offending HLO

```
%fused_computation (param_0.1: f32[], param_1.1: f32[3,1,2,2,8,8,8], param_2: f32[3,1,2,2,8,8,8])
                   -> f32[1,3,1,2,1,8,8,8] {
  %mul.45      = f32[3,1,2,2,8,8,8] multiply(%param_1.1, %param_2)
  %reduce_sum.0 = f32[3,1,2,8,8,8]   reduce(%mul.45, %param_0.1), dimensions={3}, to_apply=%region_34.55
  ROOT %bitcast.64 = f32[1,3,1,2,1,8,8,8] bitcast(%reduce_sum.0)
}
...
%ynn_fusion = f32[1,3,1,2,1,8,8,8] fusion(...), kind=kCustom, calls=%fused_computation,
              backend_config={"fusion_config":{"kind":"__ynn_fusion"}}
```

## No flag avoids it

These all leave the four `__ynn_fusion` ops in the optimized HLO and still crash:

- `--xla_cpu_use_xnnpack=false`
- `--xla_cpu_use_fusion_emitters=false`
- `--xla_cpu_experimental_ynn_fusion_type` (already empty by default)
- `--xla_cpu_experimental_xnn_graph_fusion_mode` (already `DISABLED` by default)
- `--xla_cpu_max_isa=AVX2`, `--xla_cpu_parallel_codegen_split_count=1`

The only workarounds we found are float64, jaxlib 0.10.1, or reshaping to avoid
the degenerate dimension.

## Environment

- jax 0.10.2, jaxlib 0.10.2 (broken); jax 0.10.1, jaxlib 0.10.1 (good)
- Python 3.11.15, numpy 2.4.6
- Linux x86_64, Intel Xeon W-2295 (AVX-512), 36 cores
- CPU backend (`JAX_PLATFORMS=cpu`); CUDA backend unaffected

## How we hit it

A gyrokinetics code's compressed real-FFT Poisson bracket feeds its multi-field
kernel a rank-7 operand of shape `(3, 1, 2, 1, ny, nx, nz)`; reverse mode turns
the field broadcast into exactly the reduce above. Because the crash kills the
interpreter, it takes down the whole pytest session rather than failing one test.

**Comment by rogeriojorge** [comment]
Duplicate of #39257, which reports the same `slinky::buffer<void const, 8ul>` / `make_transpose_impl` crash and predates this by a month. I've moved the additional evidence there: https://github.com/jax-ml/jax/issues/39257#issuecomment-5518786840

Short version for anyone arriving here: already fixed by google/XNNPACK@53a1797ba, which is included in XLA main's current XNNPACK pin; jaxlib 0.10.2 predates it. `JAX_ENABLE_X64=1` is a workaround, since the ynn path is f32-only.
